File: solution/01xxx/1916 - 최소비용 구하기/1916.cpp
#include <iostream>
#include <algorithm>
#include <vector>
using namespace std;

#define FASTIO cin.tie(NULL); cout.tie(NULL); ios::sync_with_stdio(false);
#define lli long long int
#define INF 2100000000
#define MAX 1111

int main() {
    FASTIO
    
    lli n, m, a, b, c;

    cin >> n;
    cin >> m;

    vector<vector<lli>> bus(n+1, vector<lli>(n+1, INF));
    vector<bool> visit(n+1, false);
    vector<lli> cost(n+1, INF);

    for(lli _ = 0; _ < m; _++) {
        cin >> a >> b >> c;
        bus[a][b] = min(bus[a][b], c);
    }

    cin >> a >> b;

    for(lli i = 1; i <= n; i++) {
        bus[i][i] = 0;
        cost[i] = bus[a][i];
    }

    visit[a] = true;

    for(lli _ = 0; _ < n; _++) {
        lli small = INF, index = 0;
        
        for(lli i = 1; i <= n; i++) {
            if(visit[i] == false && cost[i] <= small) {
                small = cost[i];
                index = i;
            }
        }

        if(index == 0) break;

        visit[index] = true;

        for(lli i = 1; i <= n; i++) {
            if(visit[i] == false) {
                if(cost[index] + bus[index][i] < cost[i]) {
                    cost[i] = cost[index] + bus[index][i];
                }
            }
        }
    }

    cout << cost[b];
}

File: solution/01xxx/1916 - 최소비용 구하기/1916.py
import sys
input = sys.stdin.readline

n = int(input())
m = int(input())

bus = [[1e9 for _ in range(n+1)] for _ in range(n+1)]

for _ in range(m):
    s, e, c = map(int, input().split())
    bus[s][e] = min(bus[s][e], c)

s, e = map(int, input().split())

cost = [1e9 for _ in range(n+1)]
visit = [False for _ in range(n+1)]

for i in range(1, n+1):
    bus[i][i] = 0
    cost[i] = bus[s][i]

visit[s] = True

for _ in range(n-2):
    small = 1e9
    next = 0

    for i in range(1, n+1):
        if visit[i] == False and cost[i] <= small:
            small = cost[i]
            next = i
    
    visit[next] = True

    for i in range(1, n+1):
        if visit[i] == False and cost[next] + bus[next][i] < cost[i]:
            cost[i] = cost[next] + bus[next][i]

print(cost[e])